src/osd/OSD.cc

```diff
@@ -4161,6 +4166,89 @@ int OSD::handle_pg_peering_evt(
   }
 }
 
+bool OSD::maybe_wait_for_max_pg(spg_t pgid, bool is_mon_create)
+{
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+
+  unsigned num_pg = 0;
+  {
+    RWLock::RLocker l(pg_map_lock);
+    num_pg = pg_map.size();
+  }
+  if (num_pg < max_pgs_per_osd) {
+    return false;
+  }
+
+  with_unique_lock(pending_creates_mutex, [=]() {
```

Comment:
should we put this inside the pg_map lock above to avoid a race condition where we compare against pg_map.size(), drop lock, another thread calls resume_creating_pg(), then we take the creating lock and add ourselves to the list?  if no other osdmap comes along it could stall.  and it looks like pending_creates_mutex is already ordered inside the pg_map lock?